cs/src/core/Index/FasterLog/FasterLog.cs

```diff
@@ -0,0 +1,321 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+#pragma warning disable 0162
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace FASTER.core
+{
+
+    /// <summary>
+    /// FASTER log
+    /// </summary>
+    public class FasterLog : IDisposable
+    {
+        private readonly BlittableAllocator<Empty, byte> allocator;
+        private readonly LightEpoch epoch;
+        private ILogCommitManager logCommitManager;
+
+        /// <summary>
+        /// Beginning address of log
+        /// </summary>
+        public long BeginAddress => allocator.BeginAddress;
+
+        /// <summary>
+        /// Tail address of log
+        /// </summary>
+        public long TailAddress => allocator.GetTailAddress();
+
+        /// <summary>
+        /// Log flushed until address
+        /// </summary>
+        public long FlushedUntilAddress => allocator.FlushedUntilAddress;
+
+        /// <summary>
+        /// Log commit until address
+        /// </summary>
+        public long CommittedUntilAddress;
+
+        /// <summary>
+        /// Create new log instance
+        /// </summary>
+        /// <param name="logSettings"></param>
+        public FasterLog(FasterLogSettings logSettings)
+        {
+            logCommitManager = logSettings.LogCommitManager ?? 
+                new LocalLogCommitManager(logSettings.LogCommitFile ??
+                logSettings.LogDevice.FileName + ".commit");
+
+            epoch = new LightEpoch();
+            allocator = new BlittableAllocator<Empty, byte>(
+                logSettings.GetLogSettings(), null, 
+                null, epoch, e => Commit(e));
+            allocator.Initialize();
+            Restore();
+        }
+
+        /// <summary>
+        /// Dispose
+        /// </summary>
+        public void Dispose()
+        {
+            allocator.Dispose();
+            epoch.Dispose();
+        }
+
+        /// <summary>
+        /// Append entry to log
+        /// </summary>
+        /// <param name="entry"></param>
+        /// <returns>Logical address of added entry</returns>
+        public unsafe long Append(Span<byte> entry)
+        {
+            epoch.Resume();
+            var length = entry.Length;
+            BlockAllocate(4 + length, out long logicalAddress);
+            var physicalAddress = allocator.GetPhysicalAddress(logicalAddress);
+            *(int*)physicalAddress = length;
+            fixed (byte* bp = &entry.GetPinnableReference())
```

Comment:
instead of pinning the entry, you can use the CopyTo API (which is also apparently faster in many cases than other memory copy APIs).

entry.CopyTo(new Span<byte>((void*)(4 + physicalAddress), length));

Reply:
I actually benchmarked the two alternatives and went with pinning as it was non-trivially faster.